### PR TITLE
[FLINK-15803][table] Update AggregateFunction and TableAggregateFunction to the new type system

### PR DESCRIPTION
**Pull request by twalthr**
## What is the purpose of the change

This updates imperative aggregate functions `AggregateFunction` and `TableAggregateFunction` to the new type system with new type inference. The new stack is activated when using `TableEnvironment.createTemporarySystemFunction` or `call(Class, ...)` in Table API. Other locations (SQL DDL, Table API Scala implicits) will be updated once we support the concept of `DataView` as well. Currently, `DataView`s are not supported.

## Brief change log

- Update the base classes `AggregateFunction` and `TableAggregateFunction`
- Enable them in `createTemporarySystemFunction`
- Update the entire stack from SQL/Table API until `AggregateUtil`
- Update some tests in different areas

## Verifying this change

This change is already covered by existing tests.

The following functions and their tests have been updated:
`CountNullNonNull, VarArgsAggFunction, Top3, WeightedAvgWithMergeAndReset`

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): no
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: yes
  - The serializers: no
  - The runtime per-record code paths (performance sensitive): no
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn/Mesos, ZooKeeper: no
  - The S3 file system connector: no

## Documentation

  - Does this pull request introduce a new feature? yes
  - If yes, how is the feature documented? JavaDocs

### CODE REVIEW

File: flink-table/flink-table-planner-blink/src/main/java/org/apache/flink/table/planner/expressions/SqlAggFunctionVisitor.java [hunk 1]
```diff
@@ -77,44 +84,72 @@ public SqlAggFunction visit(CallExpression call) {
...
-				.orElse(FunctionIdentifier.of(aggFunc.functionIdentifier()));
+		return createSqlAggFunction(
+			call.getFunctionIdentifier().orElse(null),
+			call.getFunctionDefinition());
+	}
+
+	private SqlAggFunction createSqlAggFunction(@Nullable FunctionIdentifier identifier, FunctionDefinition definition) {
+		// legacy
+		if (definition instanceof AggregateFunctionDefinition) {
+			final AggregateFunctionDefinition aggDef = (AggregateFunctionDefinition) definition;
```
Comment (aljoscha) [inline flink-table/flink-table-planner-blink/src/main/java/org/apache/flink/table/planner/expressions/SqlAggFunctionVisitor.java:104]: That's the clean code nerd in me but I think all the if-else branches could become function calls, including the new code outside the if. I.e:
```
if (definition instanceof AggregateFunctionDefinition) {
    return translateLegacyAggregateFunction(...)
} else if (...) {
    return translateLegacyTableAggregateFunction(...)
}

return translateFunction(...)
```

File: flink-table/flink-table-common/src/main/java/org/apache/flink/table/types/logical/utils/LogicalTypeChecks.java [hunk 2]
```diff
@@ -445,6 +445,15 @@ public Integer visit(StructuredType structuredType) {
 		public Integer visit(DistinctType distinctType) {
 			return distinctType.getSourceType().accept(this);
 		}
+
+		@Override
```
[the remark this answers is not in the document]
Comment (twalthr) [reply flink-table/flink-table-common/src/main/java/org/apache/flink/table/types/logical/utils/LogicalTypeChecks.java:449]: The `defaultMethod` handles all types which methods have not been overridden. See `LogicalTypeDefaultVisitor`.

File: flink-table/flink-table-common/src/main/java/org/apache/flink/table/types/utils/DataTypeUtils.java [hunk 3]
```diff
@@ -164,6 +168,33 @@ public static TableSchema expandCompositeTypeToSchema(DataType dataType) {
...
+			return dataType.getChildren();
+		}
+		return Collections.singletonList(dataType);
+	}
+
+	/**
+	 * Returns the names of the flat representation of the given data type.
+	 */
+	public static List<String> flattenToNames(DataType dataType, List<String> existingNames) {
+		final LogicalType type = dataType.getLogicalType();
```
[the remark this answers is not in the document]
Comment (twalthr) [reply flink-table/flink-table-common/src/main/java/org/apache/flink/table/types/utils/DataTypeUtils.java:188]: I'm not sure if it worth it to further combine these two methods. There are 3 branches which different logic in each branch.

File: flink-table/flink-table-planner-blink/src/main/java/org/apache/flink/table/planner/catalog/FunctionCatalogOperatorTable.java [hunk 4]
```diff
@@ -206,10 +205,13 @@ private boolean verifyFunctionKind(
 			return false;
 		}
 
-		// it would be nice to give a more meaningful exception when a scalar function is used instead
-		// of a table function and vice versa, but we can do that only once FLIP-51 is implemented
+		final FunctionKind kind = definition.getKind();
 
-		if (definition.getKind() == FunctionKind.SCALAR) {
+		if (kind == FunctionKind.TABLE) {
```
[the remark this answers is not in the document]
Comment (twalthr) [reply flink-table/flink-table-planner-blink/src/main/java/org/apache/flink/table/planner/catalog/FunctionCatalogOperatorTable.java:210]: I added a JavaDoc to this method. It basically verifies which kinds of functions are allowed to be returned from the catalog given the context information. `SqlFunctionCategory` can help for checking but is not always helpful. In any case non user-defined functions or `FunctionKind.OTHER` are definitely not supported at this location.

**Pull request closed without merging**